sky/provision/kubernetes/utils.py

```diff
@@ -1722,3 +1742,131 @@ def get_kubernetes_node_info() -> Dict[str, KubernetesNodeInfo]:
+    ]
+    return {pod.metadata.name: pod for pod in pods}
+
+
+def _remove_pod_annotations(pod: V1Pod, annotations_key: str,
+                            namespace: str) -> None:
+    """Removes specified Annotations from a Kubernetes pod."""
+    try:
+        # Remove the specified annotations
+        if pod.metadata.annotations:
+            if annotations_key in pod.metadata.annotations:
+                # Patch the pod with the updated metadata
+                body = {'metadata': {'annotations': {annotations_key: None}}}
+                kubernetes.core_api().patch_namespaced_pod(
+                    name=pod.metadata.name,
+                    namespace=namespace,
+                    body=body,
+                    _request_timeout=kubernetes.API_TIMEOUT)
+
+    except kubernetes.api_exception() as e:
+        if e.status == 404:
+            logger.warning(
+                ANNOTATIONS_POD_NOT_FOUND_ERROR_MSG.format(
+                    pod_name=pod.metadata.name,
+                    namespace=namespace,
+                    action='remove',
+                    annotations=annotations_key))
+        else:
+            with ux_utils.print_exception_no_traceback():
+                raise
+
+
+def _add_pod_annotations(pod: V1Pod, annotations: Dict[str, str],
+                         namespace: str) -> None:
+    """Adds specified Annotations on a Kubernetes pod."""
+    try:
+        # Patch the pod with the updated metadata
+        body = {'metadata': {'annotations': annotations}}
+        kubernetes.core_api().patch_namespaced_pod(
+            name=pod.metadata.name,
+            namespace=namespace,
+            body=body,
+            _request_timeout=kubernetes.API_TIMEOUT)
+
+    except kubernetes.api_exception() as e:
+        if e.status == 404:
+            logger.warning(
+                ANNOTATIONS_POD_NOT_FOUND_ERROR_MSG.format(
+                    pod_name=pod.metadata.name,
+                    namespace=namespace,
+                    action='add',
+                    annotations=annotations))
+        else:
+            with ux_utils.print_exception_no_traceback():
+                raise
+
+
+def set_autodown_annotations(handle: 'backends.CloudVmRayResourceHandle',
+                             idle_minutes_to_autostop: Optional[int],
+                             down: bool = False) -> None:
+    """Adds or removes Annotations of autodown on Kubernetes pods."""
+    annotations_keys = KUBERNETES_AUTODOWN_ANNOTATIONS_KEYS
+    tags = {
+        provision_constants.TAG_RAY_CLUSTER_NAME: handle.cluster_name_on_cloud,
+    }
+    ray_config = common_utils.read_yaml(handle.cluster_yaml)
+    provider_config = ray_config['provider']
+    namespace = get_namespace(provider_config)
+    running_pods = filter_pods(namespace, tags)
+
+    for _, pod in running_pods.items():
+        if down:
+            for annotations_key in annotations_keys:
```

Comment:
Do we need `annotations_keys` and iterating over it? Seems a little odd here.

How about removing the loop and just plainly adding two keys, something like:
```
autodown_annotations = {IDLE_MINUTES_TO_AUTOSTOP_ANNOTATION_KEY: str(idle_minutes_to_autostop),
  AUTODOWN_ANNOTATION_KEY: 'true'}
annotations.update(autodown_annotations)
```

Reply:
The purpose of the loop was to make two methods, `_add_pod_annotations` and `_remove_pod_annotations`, as general/modular as possible, which was to process each annotation one at a time and raise error per annotation separately. This would be useful in case we need to add/remove other type of Annotations in the future.

I agree that we can add all keys in annotations and call `_add_pod_annotations` just once and we can also perhaps update `_remove_pod_annotations` to process multiple annotations as well given a list of annotations keys, but both would become less generic in terms of raising errors as it will raise an error with multiple annotation names rather than pointing out the specific annotation that caused the error.

What do you think?

Reply:
Discussed offline and came to conclusion to remove the loop and run the method to add annotation one by one. Updated at [6219702](https://github.com/skypilot-org/skypilot/pull/3870/commits/6219702eeacdd216cde66ee73ce6da523c342fa0)